### ISSUE TITLE
SPI communication

### ISSUE THREAD

**Issue by malbreitenbach** [im-Kitsch/RMR_sensor_fusion#10]
We tested the SPI - communication between the Nucleo and HLP. The self-written driver on the HLP seems to work. But the builded cable is very fragile. New pre-crimped cables are already ordered with Dino's permission.
Byte transfers from Nucleo to HLP is working like expected. Byte transfers from HLP to Nucleo isn't working like expected (Nucleo just receives '0x00'). We looked at the signals with the oscilloscope and the Clock and SS is working like expected (if the clock isn't too high). The MISO and MOSI looks not clean (Up to four different voltage levels).

**Comment by malbreitenbach** [comment]
Update:
A new Cable seems to improve the communication. We did a short benchmark at different transfer speeds:

<code>

|Speed [kbs]|Nucelo's successful transfers [%]|HLP's successful transfers [%]|
| --- | --- | --- |
|357| 99,6  | 99,6 |
|7000| 87,5 | 99,1 |

</code>

**legend:** 
- speed in kilo bits per second
- successful transfer means 100 Bytes received 100% correctly

**Testing method:**
Both controllers were transmitting fix data packages (100 Bytes) with known values. At the beginning of each transmission there is a synchronization (Two - Way - Handshake). Successful and unsuccessful transmitts/receives were estimated and compared at the end.

**Results** (see EDIT)
The communication between from Nucelo->HLP is very poor. Even with very low transfer speeds the error rate is extremly high. We will build another cable to see the influence of the cable-quality. The debugging functionality of the Nucleo enables a good opportunity to see if the transfers are _really successful_ (Reading the received data manually and see if the testing algorithm works properly). On the other hand it's very hard to debug the HLP. It's not possible to set _breakpoints_ and the _Watch Expressions_ are working very poorly. We already stay in contact with the support to solve that problems and hopefully get some insight and estiminate the source of the bad error rate.

EDIT: An increase of the Nucleo's peripheral bus speed leads to much better results (see the table).

Cheers!

**Comment by raultron** [comment]
> EDIT: An increase of the Nucleo's peripheral bus speed leads to much better results (see the table).

But in the table a higher speed shows worst results, or is there an error?.

**Comment by malbreitenbach** [comment]
The speed in the table means the speed of the transfer (bits / second). That speed is different to the Bus-speed of the Nucleo. The HLP (the master) is setting the transfer speed by itself. Running with the old Nucleo's bus-speed leads to a success rate ~10%. Sorry for being so unclear. The table just shows the results for the new bus speed. I will add the old results (again), to show up the difference.

**Comment by raultron** [comment]
OK clear. 
Question: these results are using RAW data transmission right? there is no protocol involved for the moment?

**Comment by malbreitenbach** [comment]
Yes, RAW Data. The protocol will be tested in the next two days. The only protocol part which is already tested is the Two-Way-Handshake to synchronize the Nucleo and HLP.

**Comment by malbreitenbach** [comment]
@raultron 
** Performance of the protocol **
We had some test runs last firday, to see how fast the protocol works. It turned out the number of bytes in one message is not really affecting the overall performance (messages / second) while running at very high transfer speeds (>1MHz). 
We achieved an overall performance of **~330Hz** by a transfer speed of 4MHz (approx. the same performance by 7MHz). So in one second we were sending 330 messages (including >80 Bytes) of (fake*) sensory data. The protocol includes at that a function to check for the right format/length of the message and was checking the integrity of the data by a simple checksum. We are already identifying >99,9% of all erroneous messages (Of cause this number needs to be 100% at the end). 
The test is not including any sensor fusion calculations or the send back of control signals (Nucleo -> HLP). So this needs to be kept in mind. 
Currently I'm not sure how much we can improve the performance by optimizations, but the code is at this point definetly unoptimized. 
It's easy to see that there is a huge amount of overhead (besides the SPI communication) going on. The transfer itself just takes (80/4MHz=) 0.02ms which is just a very small amount of the complete 3ms (per message). In the next days I will analyze what part of the HLP - code is time critical.

(*) fake: For benchmarking reasons we used some fake data, but we already tested it successfully with true sensory data.
